File: python_backend/workspace_tools.py
import os
import subprocess
from typing import Dict, Any

WORKSPACE_ROOT = os.getcwd()

class WorkspaceTools:
    def __init__(self):
        self.workspace_root = WORKSPACE_ROOT

    def list_files(self, dir_path: str = "") -> Dict[str, Any]:
        target = os.path.join(self.workspace_root, dir_path)
        try:
            if not os.path.exists(target):
                return {"error": f"Path not found: {dir_path}"}
            
            items = []
            for entry in os.scandir(target):
                items.append({
                    "name": entry.name,
                    "isDirectory": entry.is_dir(),
                    "size": entry.stat().st_size if entry.is_file() else None
                })
            return {"path": dir_path or "/", "items": items}
        except Exception as e:
            return {"error": str(e)}

    def read_file(self, file_path: str) -> Dict[str, Any]:
        target = os.path.join(self.workspace_root, file_path)
        try:
            if not os.path.exists(target):
                return {"error": f"File not found: {file_path}"}
            with open(target, "r", encoding="utf-8", errors="ignore") as f:
                content = f.read(10000)
            return {"filePath": file_path, "content": content}
        except Exception as e:
            return {"error": str(e)}

    def write_file(self, file_path: str, content: str) -> Dict[str, Any]:
        target = os.path.join(self.workspace_root, file_path)
        try:
            parent = os.path.dirname(target)
            if not os.path.exists(parent):
                os.makedirs(parent, exist_ok=True)
            with open(target, "w", encoding="utf-8") as f:
                f.write(content)
            return {"success": True, "filePath": file_path, "bytesWritten": len(content)}
        except Exception as e:
            return {"error": str(e)}

    def execute_command(self, command: str) -> Dict[str, Any]:
        try:
            if "rm -rf /" in command or "format c:" in command:
                return {"error": "Command blocked by security policy"}
            
            result = subprocess.run(
                command,
                shell=True,
                cwd=self.workspace_root,
                capture_output=True,
                text=True,
                timeout=30
            )
            return {
                "success": result.returncode == 0,
                "stdout": result.stdout[:4000],
                "stderr": result.stderr[:1000]
            }
        except Exception as e:
            return {"success": False, "error": str(e)}

workspace_tools = WorkspaceTools()

File: server/src/workspace_tools.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

const WORKSPACE_ROOT = process.cwd();

/**
 * Real Workspace Tool Execution Handler for AI Agents
 */
class WorkspaceTools {
  constructor() {
    this.workspaceRoot = WORKSPACE_ROOT;
  }

  // List files in workspace
  listWorkspaceFiles(dirPath = '') {
    const targetPath = path.join(this.workspaceRoot, dirPath);
    try {
      if (!fs.existsSync(targetPath)) return { error: `Path does not exist: ${dirPath}` };
      const items = fs.readdirSync(targetPath, { withFileTypes: true });
      return {
        path: dirPath || '/',
        items: items.map(item => ({
          name: item.name,
          isDirectory: item.isDirectory(),
          size: item.isFile() ? fs.statSync(path.join(targetPath, item.name)).size : null
        }))
      };
    } catch (err) {
      return { error: err.message };
    }
  }

  // Read file content
  readFileContent(filePath) {
    const targetPath = path.join(this.workspaceRoot, filePath);
    try {
      if (!fs.existsSync(targetPath)) return { error: `File not found: ${filePath}` };
      const content = fs.readFileSync(targetPath, 'utf8');
      return { filePath, content: content.slice(0, 10000) }; // cap at 10k chars
    } catch (err) {
      return { error: err.message };
    }
  }

  // Write content to file
  writeFileContent(filePath, content) {
    const targetPath = path.join(this.workspaceRoot, filePath);
    try {
      const dir = path.dirname(targetPath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.writeFileSync(targetPath, content, 'utf8');
      return { success: true, filePath, bytesWritten: Buffer.byteLength(content) };
    } catch (err) {
      return { error: err.message };
    }
  }

  // Execute terminal command safely inside workspace
  executeCommand(command) {
    return new Promise((resolve) => {
      // Basic sanity check to prevent dangerous system commands
      if (command.includes('rm -rf /') || command.includes('format c:')) {
        return resolve({ error: 'Command blocked by security policy' });
      }

      exec(command, { cwd: this.workspaceRoot, timeout: 30000 }, (error, stdout, stderr) => {
        if (error) {
          return resolve({
            success: false,
            error: error.message,
            stderr: stderr.slice(0, 2000),
            stdout: stdout.slice(0, 2000)
          });
        }
        resolve({
          success: true,
          stdout: stdout.slice(0, 4000),
          stderr: stderr.slice(0, 1000)
        });
      });
    });
  }
}

module.exports = new WorkspaceTools();
